fetch_worker.go

```diff
@@ -266,3 +284,136 @@ func (w *fetchWorker) buildFetchData(
+	return w.config.table
+}
+
+func (w *fetchWorkerByLimit) countTotal(ctx context.Context) (uint64, error) {
+	var total struct {
+		Total uint64 `db:"total"`
+	}
+
+	query, args, err := squirrel.Select("COUNT(*) as total").From(w.config.table).ToSql()
+	if err != nil {
+		return 0, fmt.Errorf("failed to build query: %w", err)
+	}
+
+	row := w.db.QueryRowxContext(ctx, query, args...)
+	if err := row.StructScan(&total); err != nil {
+		return 0, fmt.Errorf("failed to fetch total from %s: %w", w.config.table, err)
+	} else if err := row.Err(); err != nil {
+		return 0, fmt.Errorf("failed to fetch total from %s: %w", w.config.table, err)
+	}
+
+	sdk.Logger(ctx).Debug().Str("query", query).Any("args", args).Msg("count query")
+
+	return total.Total, nil
+}
+
+func (w *fetchWorkerByLimit) fetchStartEnd(ctx context.Context) (isTableEmpty bool, err error) {
+	total, err := w.countTotal(ctx)
+	if err != nil {
+		return false, fmt.Errorf("failed to get total: %w", err)
+	} else if total == 0 {
+		return true, nil
+	}
+
+	w.end = total
+
+	return false, nil
+}
+
+func (w *fetchWorkerByLimit) run(ctx context.Context) (err error) {
+	tx, err := w.db.BeginTxx(ctx, &sql.TxOptions{
+		Isolation: sql.LevelRepeatableRead,
+		ReadOnly:  true,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to create transaction: %w", err)
+	}
+	defer func() {
+		if err = tx.Commit(); err != nil {
+			err = fmt.Errorf("failed to commit transaction: %w", err)
+		}
+	}()
+
+	sdk.Logger(ctx).Debug().Msgf("obtained tx for table %v", w.config.table)
+
+	for offset := uint64(0); offset < w.end; offset += w.config.fetchSize {
+		query, args, err := squirrel.
+			Select("*").From(w.config.table).
+			Limit(w.config.fetchSize).Offset(offset).ToSql()
+		if err != nil {
+			return fmt.Errorf("failed to build query: %w", err)
+		}
+
+		sdk.Logger(ctx).Debug().Str("query", query).Any("args", args).Msg("created query")
+
+		rows, err := tx.QueryxContext(ctx, query, args...)
+		if err != nil {
+			return fmt.Errorf("failed to query rows: %w", err)
+		}
+		defer rows.Close()
+
+		for rows.Next() {
+			row := opencdc.StructuredData{}
+			if err := rows.MapScan(row); err != nil {
+				return fmt.Errorf("failed to map scan row: %w", err)
+			}
+
+			// convert the values so that they can be easily serialized
+			for key, val := range row {
+				row[key] = common.FormatValue(val)
+			}
+
+			keyBytes, err := json.Marshal(row)
+			if err != nil {
+				return fmt.Errorf("failed to marshal row: %w", err)
+			}
+			encodedKey := base64.StdEncoding.EncodeToString(keyBytes)
+
+			// we don't really have any way to get the key from the table, so we
+			// make up for one
+			key := snapshotKey{
+				Key:   "",
+				Value: encodedKey,
+			}
+
+			w.data <- fetchData{
+				key:     key,
+				table:   w.config.table,
+				payload: row,
+
+				// ignoring position, we start from 0 each time
+				// TODO: we could save the offset as a position, we might want
+				// to do it in the future.
```

Comment:
Interesting thought. We probably cannot get any additional guarantees about the data (we could still either miss data or get duplicates). So I  think that for now, it's safer to get duplicates than to miss data. Duplicates could cause errors in the destination, because of unique constraints, but hey, it's called an unsafe snapshot anyway.:)